nix_2_3: drop and raise minver to 2.18

wolfgangwalther opened the pull request:
Nix 2.3 was marked insecure in #420974.

In #427437, we updated the pinned nixpkgs for CI at which point we were not able to test Nix 2.3 in CI anymore and had to remove it from the `parse` check.

This brought up https://github.com/NixOS/nixpkgs/pull/427437#issuecomment-3102652854 and https://github.com/NixOS/nixpkgs/pull/427437#issuecomment-3112985078. The same questions asked therein, are still open: What's the plan for Nix 2.3 going forward?

I have neither the ability, nor the intent to backport the patches myself, "removal" is the only suggestion I can practically make, thus this PR.

## Things done

- Nixpkgs Release Notes
  - [x] Package update: when the change is major or breaking.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wolfgangwalther committed:
nix_2_3: drop

This has been marked insecure a while ago, as some CVEs have not been
backported. Even if *some* CVEs are fixed, we'd need **all** of them to
be, to get it back into the cache.

Not having it in the cache means, we can not test it in CI. This means
we can't make sure to actually support this version to evaluate Nixpkgs.

emilazy commented:
2.18 as a baseline was what was previously discussed (with you and @Ma27, I believe?). It seems reasonable to me. I don’t know if we want to set a particular point version or not; I believe we’re on 2.3.bignumber because a bunch of things got backported specifically so they could be used in Nixpkgs.

It’s unfortunate that 2.18 is no longer in Nixpkgs so we can’t use it directly for eval checks, but since we check parsing against Lix already it seems sensible to me to use it for baseline eval checks too, and in any case having a CI baseline that gets security patches seems like a good idea. Maybe we can move `nixVersions.minimal` outside of `nixVersions` to solve the jurisdictional issues here and call it `nixForEvalBaseline` or something.

I know @flokli said he still considers 2.3 evaluation compatibility important the last time we talked about it on Matrix. That was before the vulnerabilities, though. If Tvix/Snix match 2.3, I suppose we could add those to CI? But IIRC they can’t evaluate all of Nixpkgs properly yet, so we might not want to make it load‐bearing. I could be wrong though.

The PR title is inaccurate.

RaitoBezarius commented:
> My understanding is that Lix do not intend to introduce any new language‐accessible functionality without explicit feature gating or versioning – at least out of “obviously unstable” edge cases? – but I can’t speak with confidence there; maybe that only applies to syntax rather than built‐ins? Hopefully someone from the Lix team can comment more authoritatively. But there are definitely things in newer versions of Nix that Lix does not implement and I think does not intend to, so I think it’s sensible as a “keep us honest” choice for 2.18 compatibility.

Lix cares about being hash equivalent to Nix 2.18 (i.e. drvs equivalence, outpath equivalence, etc.) & cares about language stability including with respect to laziness behaviors. All experimental features are out of scope, obviously.

Anything else is a bug.

emilazy commented:
> I wonder whether we should do all our Eval with the version that is closest to the minimum, aka 2.3 now or 2.18 after this PR. And then only do the additional parse checks with latest versions.

I think it’s good to ensure all of (minimum versions, default versions, latest versions) work. That would be too much, though, so we have to pick and choose. I don’t have a strong opinion on what we should pick if we can only have one except that it seems best to be conservative about unnecessary changes in this PR since it’s already likely to be a matter of some controversy. My hope is that merge queues would allow us to cover at least both the minimum version and the default or latest, by punting all but one eval into the queue job; we could then also ensure hash stability between versions, which is important for Nixpkgs integrity and also I believe something that has regressed in the past (due to implementation bugs rather than Nixpkgs ones).

(Modulo available CI resources, checking eval against Lix seems like a sensible idea in general given a lot of contributors and deployments use it, even aside from how we encode the minimum version stuff.)

emilazy commented on pkgs/tools/package-management/nix/default.nix:
```diff
@@ -234,21 +208,9 @@ lib.makeExtensible (
-          nix;
+      # Lix has forked from Nix 2.18 and exposes its `builtins.nixVersion` as "2.18.3-lix".
+      # Thus, this is technically the minimum supported nix version.
+      minimum = pkgs.lix;
```
I think that since the Nix team’s policy is to only support the latest version and the default version(s?) in Nixpkgs, `minimum` doesn’t make much sense at this point. But I do think putting 2.18 in `lib/minver.nix` makes sense as a “language, cache, etc. baseline” in lieu of a better versioning scheme for those. Maybe having a specific package alias for the baseline doesn’t make sense, though, if it’ll only be consumed by CI that can manually select what to use. That means `minver.nix` could drift from reality, but in the absence of actual Nix 2.18 it’s already a little fuzzy.

wolfgangwalther committed:
nixVersions.minimum: drop

The concept of this alias becomes questionable once we move past 2.18,
where Lix was forked. We should probably move to a feature-detection
based approach for lib/minver.nix eventually, too.

emilazy commented:
> If we just compare minimum vs latest, then:
> 
> * If we check only latest in Nixpkgs, we risk using features that older versions don't understand.
> * If we check only minimum in Nixpkgs, we can still rely on both Nix and Lix having a strong incentive to be able to evaluate older code.

I think we can expect newer versions to generally have better performance, though, which is important for CI. (Though I know both Nix and Lix have had regressions on this front, but I believe the recent trend is upwards for both.)

> My current gut feel is going towards: We should define both a minimum Nix and a minimum Lix version and test against both of these. But only these, not stable or latest - that would use too many resources.

I feel like we should certainly test against default, right? Default is what we actually ship. Non‐default latest versions are less relevant by any metric.

MattSturgeon commented:
Without wanting to detail this discussion too much, do we know why 2.18 was dropped?

It'd be nice to have it back if we want to treat it as our minimum version.

As for 2.3, that will be 6 years old in September. I think it is ok to bump our "minimum" if there's a strong reason (like CVEs).

---

I agree in principle that it is more important to run CI on minimum versions, because that is where issues are most likely to be missed by local testing.

Ideally we would test on all versions, but beyond the basic "parse" tests that'd be way too expensive.

flokli commented:
> I know @flokli said he still considers 2.3 evaluation compatibility important the last time we talked about it on Matrix. That was before the vulnerabilities, though. If Tvix/Snix match 2.3, I suppose we could add those to CI? But IIRC they can’t evaluate all of Nixpkgs properly yet, so we might not want to make it load‐bearing. I could be wrong though.

Sorry for the long delay with reacting to this as all. I had a few IRL things in the last few weeks and am still lagging behind.

IMHO it's important to be able evaluate nixpkgs with Nix 2.3, and ensure we come up with the same .drv hashes [^1]. This is to ensure we don't accidentially break hashing / eval semantics. Which unfortunately happened a few times, and got surfaced by impls like Tvix/Snix, on a large testbed such as nixpkgs.

I don't use and don't think you should use Nix 2.3 on a "real" NixOS system, in daily-to-day use.

So it having known vulnerabilities (as long as they're marked as such) is fine, if it's only used as an evaluation engine / reference implementation, to ensure we stay stable with this. None of the known vulns are applicable at eval time.

I don't think Nix 2.3 usage as a builder is something we should do. But I also don't think the things people are looking forward to be able to do after bumping minver would actually change hashes between 2.3 vs 2.15.

A stdenv without the `$NIX_ATTRS_SH_FILE` shouldn't have a different drv hash when evaluated on 2.3 vs 2.15.

> The aws-sdk-cpp-old dependency of nix_2_3 is now marked broken [12db043 (#428410)](https://github.com/NixOS/nixpkgs/pull/428410/commits/12db043825cb2574093f8240b859a33ee1f0d143)

Nix 2.3 without AWS support should still evaluate derivations in nixpkgs the same as with 2.3.

--

So if we see Nix 2.3 simply as a "evaluation reference", discourage 2.3 usage on live systems by piping in a warn, and maybe build without AWS support to make the expression easier to maintain, what's the remaining language features that bumping minver would give us? Is it only path interpolation support?

--

[^1]: and yes, we absolutely should also do the same "ensure it evaluates the same" checking with Lix, as well as with Nix 2.15. Even if Nix upstream doesn't want to "support" it, if nixpkgs is promising to stay compatible with its language version, we need to test that this doesn't regress.

MattSturgeon commented:
I agree it makes sense to think of this as a reference implementation for checking eval compatibility (etc) without actually caring about CVEs in CI.

Would it be reasonable to drop nix 2.3 from `pkgs` and move the package into the `ci/` subtree? If the nix team are only wanting to maintain default and latest, then this probably makes sense.

Or if we wish to keep it in pkgs for now, it could be deprecated and/or mark insecure there, but CI usages could override or directly `callPackage` the package, _without_ marking it as insecure.

sternenseemann commented:
> it could be deprecated and/or mark insecure there

It has been marked as insecure in https://github.com/NixOS/nixpkgs/pull/420974

piegamesde commented:
> So if we see Nix 2.3 simply as a "evaluation reference", discourage 2.3 usage on live systems by piping in a warn, and maybe build without AWS support to make the expression easier to maintain, what's the remaining language features that bumping minver would give us? Is it only path interpolation support?

For now this is an acceptable solution, but long term the Nix language absolutely needs to evolve, and at the very least this requires the ability to add new features to the language. We'll need a long-term plan for checking compatibility once that happens

flokli commented:
The "if it's marked as insecure, it's not built by Hydra, and by this we cannot use it in CI" argument is a bit too absolute. It's not that all GH actions pipelines need to use c.n.o.

You could also have a static build, pull it from an older nixpkgs checkout, or pull it from another binary cache.

Mic92 commented:
Here some traffic statistics from cache.nixos.org to drive the decision:

```
============================================================
NIX AND LIX VERSION ANALYSIS - Last Week (July 29 - Aug 4, 2025)
============================================================

Total requests: 16,063,644
Sample: 54 files (8 time points per day × 7 days)

Overall Usage Summary:
--------------------------------------------------
Nix: 15,740,409 ( 97.99%)
Lix: 323,235 (  2.01%)

Top 30 Nix major.minor versions:
--------------------------------------------------
 1. Nix/2.28     3,417,424 (21.27%)
 2. Nix/2.30     3,120,911 (19.43%)
 3. Nix/2.18     2,699,252 (16.80%)
 4. Nix/2.24     1,319,292 ( 8.21%)
 5. Nix/2.15     1,233,324 ( 7.68%)
 6. Nix/2.26     1,033,342 ( 6.43%)
 7. Nix/2.29      726,384 ( 4.52%)
 8. Nix/2.17      510,654 ( 3.18%)
 9. Nix/2.23      361,902 ( 2.25%)
10. Nix/2.20      296,762 ( 1.85%)
11. Nix/2.13      212,316 ( 1.32%)
12. Nix/2.22      205,543 ( 1.28%)
13. Nix/2.19      200,031 ( 1.25%)
14. Nix/2.3        92,768 ( 0.58%)
15. Nix/2.25       77,048 ( 0.48%)
16. Nix/2.16       60,340 ( 0.38%)
17. Nix/2.21       41,458 ( 0.26%)
18. Nix/2.8        35,598 ( 0.22%)
19. Nix/2.27       30,589 ( 0.19%)
20. Nix/2.31       27,551 ( 0.17%)
21. Nix/2.11       11,938 ( 0.07%)
22. Nix/2.10        6,562 ( 0.04%)
23. Nix/2.12        4,893 ( 0.03%)
24. Nix/2.14        3,873 ( 0.02%)
25. Nix/2.5         3,179 ( 0.02%)
26. Nix/2.6         2,812 ( 0.02%)
27. Nix/2.9         1,772 ( 0.01%)
28. Nix/2.4         1,229 ( 0.01%)
29. Nix/2.0           833 ( 0.01%)
30. Nix/2.2           460 ( 0.00%)

Top 30 Lix major.minor versions:
--------------------------------------------------
 1. Lix/2.93      179,011 ( 1.11%)
 2. Lix/2.91       95,132 ( 0.59%)
 3. Lix/2.94       33,311 ( 0.21%)
 4. Lix/2.92       15,428 ( 0.10%)
 5. Lix/2.90          353 ( 0.00%)
```

Source is here: https://github.com/Mic92/nix-version-analysis

sternenseemann commented:
> Given that all of these must be on older releases of Nixpkgs, where these versions are still available, we can assume that a fair share of this 0.58% is _also_ on older Nixpkgs revisions

Usage numbers for Nix versions that have been removed from Nixpkgs don't really say anything about what Nixpkgs revision users of Nix versions that haven't been removed are on.

Mic92 commented:
> > long term the Nix language absolutely needs to evolve
> 
> This is frankly unrelated to this discussion since no one has (thus far) been able to establish a consensus on how to evolve the Nix language in the future. Evolution isn't prevented by Nix 2.3 or 2.18 being the min version for Nixpkgs, but a way to do it without waiting until support for Nix versions without support for it can safely be dropped.
> 
> The only changes that have been causing breakage has been the introduction of path interpolation which Nix < 2.4 can't parse. This is a new language feature that saves you typing a few characters and isn't even used very often (single digit amount of instances per year I'd guess). There are some new builtins which can be polyfilled except `ceil` and `floor` which aren't really useful in Nixpkgs.

There is also unsafeDiscardReferences: https://github.com/NixOS/nixpkgs/pull/425700

MattSturgeon commented:
To answer the core question: yes, I think we can drop 2.3 from `pkgs`.

Currently we aren't running the minimum version eval test, because the package is insecure; so dropping the package doesn't make that situation any worse.

We _should_ try to find another way to run the minimum version eval test, though. Otherwise eval compatibility regressions will surely start creeping in.

---

I don't think it is nixpkgs responsibility to check hash consistency between versions, but I'm not opposed to doing that in CI. _Ideally_ the nix (+lix) repo(s) would handle that with their own regression tests.

IMO, the only reason we should test the minimum version is to enforce eval compatibility. I.e. to ensure we don't inadvertently start using new builtins, language features, etc; without first deliberately bumping the minimum version.

I agree we don't need nix 2.3 in `pkgs` for this, but we do need a cached build _somewhere_. Maybe it makes sense for CI to have its own dedicated binary cache for things like this (e.g. a cachix or another hydra instance)?

Alternatively a quick fix is to pull 2.3 from an old nixpkgs release that's still cached by hydra. But that would mean fetching and evaluating _another_ copy of nixpkgs during CI runs.

MattSturgeon commented:
> I'm currently implementing that "ratchet style" check for eval consistency between nix and lix versions.

Is the goal to check nix implementations eval our tests correctly, or to check that nixpkgs itself is compatible with the minimum-supported version? (Or both)?

Having a check for eval consistency between nix versions/implementations sounds useful, but my understanding was the reason we kept the minimun-version test around was to ensure nixpkgs doesn't inadvertently start using newer syntax or language features.

So even with this new "ratchet" test, I still think we need to find a way to package and cache the minimum supported version, ideally outside of the `pkgs` tree.

> I get a different hash for `tests.devShellTools.nixos` (and `tests.devShellTools.unstructuredDerivationInputEnv`)

That's concerning...

MattSturgeon commented:
> > So even with this new "ratchet" test, I still think we need to find a way to package and cache the minimum supported version, ideally outside of the `pkgs` tree.
> 
> Well, no, not really. The "minimum" you're referring to would be 2.18 after this PR. But 2.18 shouldn't be thought of a "minimum version" anymore, imho. "Exposing a version of _at least_ 2.18" is **one feature** that each supported implementation needs to "support". We can extend this to other feature checks later. But we can't "raise" this version, because Lix is just not going to report anything higher, ever.
> 
> This does _not_ mean, that Nix 2.18 is factually still supported after this PR. It would not be, because it's not available in Nixpkgs anymore.

I'm not convinced yet.

Currently this PR _is_ titled "drop 2.3 and raise minver to 2.18", however we also discussed the possibility of dropping 2.3 (from pkgs) _without_ raising the minver.

Either way, we need the minver to be packaged somewhere (even somewhere out-of-tree) to validate that that version actually evaluates nixpkgs. Regardless of whether it is 2.3 or 2.18.

IIUC, the reason the minver doesn't need to be supported/packaged-in this version of nixpkgs is because we want nixpkgs to be able to be evaluated by nix implementations installed on older systems. E.g. imagine someone with a NixOS 24.11 install wants to update to a new version, even install one or two packages from the latest unstable. Their `nix` is installed from 24.11, and may now be "unsupported", but they still need to be able to evaluate the latest nixpkgs revision (currently 25.11). I.e. minver doesn't relate to support of that nix version, but eval compatibility with it.

sternenseemann commented:
> `--option eval-system`, thus we'd have to run this eval on 4 different GitHub actions runners instead of `ubuntu-24.04-arm`. This is not going to work well, especially for darwin.

Why are we setting `eval-system`? Nixpkgs can be called purely and I'd say any place that uses `builtins.currentSystem` instead of the appropriate  `localSystem`/`crossSystem`/`*Platform` is arguably a bug.

MattSturgeon commented on ci/default.nix:
```diff
@@ -110,8 +104,7 @@ rec {
     lix = pkgs.callPackage ./parse.nix { nix = pkgs.lix; };
-    # TODO: Raise nixVersions.minimum to 2.24 and flip back to it.
-    minimum = pkgs.callPackage ./parse.nix { nix = pkgs.nixVersions.nix_2_24; };
+    nix_2_24 = pkgs.callPackage ./parse.nix { nix = pkgs.nixVersions.nix_2_24; };
```
Could we have a "TODO test the minimum version (2.18)" style comment?

Ideally referencing a relevant comment on this PR or a tracking issue.

MattSturgeon commented on doc/release-notes/rl-2511.section.md:
```diff
@@ -14,6 +14,9 @@
 
 <!-- To avoid merge conflicts, consider adding your item at an arbitrary place in the list instead. -->
 
+- `nixVersions.nix_2_3` has been dropped because it was insecure and unmaintained.
+  The minimum version of Nix required to evaluate Nixpkgs has been raised from 2.3 to 2.18.
```
This should probably be two separate release note entries.

MattSturgeon commented on lib/tests/release.nix:
```diff
@@ -2,21 +2,14 @@
   pkgsBB ? pkgs.pkgsBuildBuild,
   nix ? pkgs-nixVersions.stable,
   nixVersions ? [
-    pkgs-nixVersions.minimum
+    pkgs-nixVersions.nix_2_24
     nix
     pkgs-nixVersions.latest
   ],
```
And/or a TODO comment here for testing the minver?

MattSturgeon commented on pkgs/by-name/ni/nixos-rebuild-ng/package.nix:
```diff
@@ -121,9 +121,9 @@ python3Packages.buildPythonApplication rec {
-        with_nix_2_3 = nixos-rebuild-ng.override {
-          # oldest / minimum supported version in nixpkgs
-          nix = nixVersions.nix_2_3;
+        with_nix_2_24 = nixos-rebuild-ng.override {
+          # oldest supported version in nixpkgs
+          nix = nixVersions.nix_2_24;
```
As these are passthru tests, it feels reasonable to add back `with_minimum` at some point. Should we have a TODO here?

MattSturgeon commented on pkgs/tools/package-management/nix/default.nix:
```diff
@@ -233,23 +207,6 @@ lib.makeExtensible (
 
       latest = self.nix_2_29;
 
-      # The minimum Nix version supported by Nixpkgs
-      # Note that some functionality *might* have been backported into this Nix version,
-      # making this package an inaccurate representation of what features are available
-      # in the actual lowest minver.nix *patch* version.
-      minimum =
-        let
-          minver = import ../../../../lib/minver.nix;
-          major = lib.versions.major minver;
-          minor = lib.versions.minor minver;
-          attribute = "nix_${major}_${minor}";
-          nix = self.${attribute};
-        in
-        if !self ? ${attribute} then
-          throw "The minimum supported Nix version is ${minver} (declared in lib/minver.nix), but pkgs.nixVersions.${attribute} does not exist."
-        else
-          nix;
-
```
Should this alias to 2.24 as the oldest supported (and packaged) version?

It feels slightly out-of-place for us to remove the `nixVersion.miniumum` attribute entirely, without input from the nix team.

That said, if the nix team are ok with removing the attribute, then I'm fine with it.

MattSturgeon reviewed (commented):
> We can then still discuss in a follow up:
> * Do we want to keep or replace the concept of "minver"?
> * If we keep "minver": Do we need to Eval _with the exact minimum version_, at least sometimes?


Yes, that SGTM. My earlier feedback wasn't intended to be actioned in this PR, but this PR felt like the most relevant place to have the discussion.

 - I'm strongly in favour of removing the unmaintained `nix_2_3` package.
 - I'm neutral to bumping the minver.
 - I'd like the minver to get some eval testing, but in a separate PR.

Nits and discussion below:

Mic92 commented:
> > I have **not** compared full nixpkgs to Nix 2.3, yet.
> 
> I have tried to do this now: It's not easily possible, if at all.
> 
> Nix 2.3's nix-env / nix-instantiate does not support:
> 
>     * `--option eval-system`, thus we'd have to run this eval on 4 different GitHub actions runners instead of `ubuntu-24.04-arm`. This is not going to work well, especially for darwin.
> 
>     * `--out-path` + `--json` - that was only implemented in 2.6: [NixOS/nix@6525761](https://github.com/NixOS/nix/commit/65257614ea77437a75447c3b5936f240e581ede9) - we'd need to rewrite the script to not produce and consume JSON anymore.
> 
>     * `--option restrict-eval true` + `-I <path-to-nixpkgs>` behaves differently and just throws an error the way we use it for the attrpaths step right now. We could remove this for an initial verification, but not keep it that way, I think.
> 
> 
> It would require significant changes to our Eval workflow to even be _able_ to run it with Nix 2.3. And then we'd still need to get it from somewhere, if it's not available from within the pinned nixpkgs we have.
> 
> This does not mean that Nixpkgs actually evals differently, far from it - but the tooling we have just doesn't allow checking it.
> 
> With the exceptions mentioned in [#428076 (comment)](https://github.com/NixOS/nixpkgs/pull/428076#issuecomment-3169049246) Nixpkgs currently still evaluates the same with the different Lix and Nix versions available in `pkgs`. We can keep validating this with the ratchet-style workflow that I will open a PR for soon. (Edit: #427724)
> 
> TLDR for this PR:
> 
>     * Nix 2.3 is not supposed to be used by users, thus the package itself can be removed.
> 
>     * We can only validate Nixpkgs evaluating properly with Nix 2.3 with a lot of work for the various reasons outlined above and earlier in this PR.
> 
>     * Raising minver to 2.18 should not be a practical problem for any upgrades: NixOS 23.11 was already on Nix 2.18 by default. If somebody was still running something older and would like to update, they can do that step by step.
> 
>     * The theoretical concern whether we are actually testing minver or not is not worsened by this PR at all, in fact the opposite is true: The gap between what we currently test and what minver is at, is narrowing with this PR.
> 
> 
> At this point we should just call it a day and merge this PR.
> 
> We can then still discuss in a follow up:
> 
>     * Do we want to keep or replace the concept of "minver"?
> 
>     * If we keep "minver": Do we need to Eval _with the exact minimum version_, at least sometimes?
> 
> 
> **Barring objections, I'd like to merge this PR within the next week.** If you, the reader, would like to object, feel free to downvote this comment for quick feedback, but please eventually elaborate on:
> 
>     * What exactly you're objecting to and
> 
>     * What you're proposing to do instead and
> 
>     * Whether you'd be ready to implement your proposal yourself or not.

Sounds good.

MattSturgeon commented on doc/release-notes/rl-2511.section.md:
```diff
@@ -14,6 +14,10 @@
 
 <!-- To avoid merge conflicts, consider adding your item at an arbitrary place in the list instead. -->
 
+- `nixVersions.nix_2_3` has been dropped because it was insecure and unmaintained.
```
Should this also mention `nixVersions.minimum`, which has also been removed?

On the one hand, aliases don't usually need to be mentioned in release notes. On the other hand, this alias is more like a meta package that isn't tightly coupled to 2.3.


```suggestion
- `nixVersions.nix_2_3` has been dropped because it was insecure and unmaintained.
  Its `nixVersions.minimum` alias has also been dropped for the same reason.
```


```suggestion
- `nixVersions.nix_2_3` and `nixVersions.minimum` have been dropped because they were insecure and unmaintained.
```

?

MattSturgeon reviewed (approved):
LGTM

Mic92 commented:
Also extracted pythonix commits into its own PR: https://github.com/NixOS/nixpkgs/pull/433036
